modules/fundamental/src/purl/service/mod.rs

```diff
@@ -230,41 +231,84 @@ impl PurlService {
         }
     }
 
-    pub async fn purl_by_purl<C: ConnectionTrait>(
+    #[instrument(skip(self, connection), err(level=tracing::Level::INFO))]
+    pub async fn fetch_purl_details<C: ConnectionTrait, I: AsRef<str> + Debug>(
         &self,
-        purl: &Purl,
-        deprecation: Deprecation,
+        identifiers: &[I],
+        deprecated: Deprecation,
         connection: &C,
-    ) -> Result<Option<PurlDetails>, Error> {
-        let canonical = CanonicalPurl::from(purl.clone());
-        match qualified_purl::Entity::find()
-            .filter(qualified_purl::Column::Purl.eq(canonical))
-            .one(connection)
+    ) -> Result<HashMap<String, PurlDetails>, Error> {
+        let (purls, uuids): (Vec<_>, Vec<_>) = identifiers
+            .iter()
+            .partition(|key| key.as_ref().starts_with("pkg:"));
+
+        let purls: Vec<_> = purls
+            .iter()
+            .map(|k| Purl::from_str(k.as_ref()).map_err(Error::Purl))
+            .collect::<Result<_, _>>()?;
+
+        let uuids: Vec<_> = uuids
+            .iter()
+            .map(|k| Uuid::from_str(k.as_ref()).map_err(Error::Uuid))
+            .collect::<Result<_, _>>()?;
+
+        let details = self
+            .purls_by_purl(&purls, deprecated, connection)
             .await?
-        {
-            Some(purl) => Ok(Some(
-                PurlDetails::from_entity(None, None, &purl, deprecation, connection).await?,
-            )),
-            None => Ok(None),
+            .into_iter()
+            .map(|detail| (detail.head.purl.to_string(), detail))
+            .chain(
```

Comment:
I liked your use of `chain` here, even though I refactored it a bit.

Reply:
Thanks!